src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesCodeFixProvider.cs

```diff
@@ -0,0 +1,702 @@
+
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(IDEDiagnosticIds.ExpressionValueIsUnusedDiagnosticId,
+                                                                                                    IDEDiagnosticIds.ValueAssignedIsUnusedDiagnosticId);
+
+        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var diagnostic = context.Diagnostics[0];
+            var preference = AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.GetUnusedValuePreference(diagnostic);
+            var isRemovableAssignment = AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.GetIsRemovableAssignmentDiagnostic(diagnostic);
+
+            string title;
+            if (isRemovableAssignment)
+            {
+                // Recommend removing the redundant constant value assignment.
+                title = FeaturesResources.Remove_redundant_assignment;
+            }
+            else
+            {
+                // Recommend using discard/unused local for redundant non-constant assignment.
+                switch (preference)
+                {
+                    case UnusedValuePreference.DiscardVariable:
+                        if (IsForEachIterationVariableDiagnostic(diagnostic, context.Document))
+                        {
+                            // Do not offer a fix to replace unused foreach iteration variable with discard.
+                            // User should probably replace it with a for loop based on the collection length.
+                            return Task.CompletedTask;
+                        }
+
+                        title = FeaturesResources.Use_discard_underscore;
+                        break;
+
+                    case UnusedValuePreference.UnusedLocalVariable:
+                        title = FeaturesResources.Use_discarded_local;
+                        break;
+
+                    default:
+                        return Task.CompletedTask;
+                }
+            }
+
+            context.RegisterCodeFix(
+                new MyCodeAction(
+                    title,
+                    c => FixAsync(context.Document, diagnostic, context.CancellationToken),
+                    equivalenceKey: GetEquivalenceKey(preference, isRemovableAssignment)),
+                diagnostic);
+
+            return Task.CompletedTask;
+        }
+
+        private static bool IsForEachIterationVariableDiagnostic(Diagnostic diagnostic, Document document)
+        {
+            // Do not offer a fix to replace unused foreach iteration variable with discard.
+            // User should probably replace it with a for loop based on the collection length.
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var root = diagnostic.Location.SourceTree.GetRoot();
+            return syntaxFacts.IsForEachStatement(root.FindNode(diagnostic.Location.SourceSpan));
```

Comment:
you can just get the node by call diagnostic.Location.FindNode (or something like that).  No need for the root here.